text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+│ │   ┃                       ┃                                    │ │
+│ │   ┃ const ValueSized      ┃ ValueSized                         │ │
+│ │   ┃ {type, target, value} ┃ {type, target, runtime env, value} │ │
+│ │   ┗━━━━━━━━━━━━━━━━━━━━━━━┛                                    │ │
+│ └────────────────────────────────────────────────────────────────┘ │
+│  Pointee                                                           │
+│  {*}                                                               │
+└────────────────────────────────────────────────────────────────────┘
+```
+
+`const Sized` is implemented on types which require knowledge of only the
+type and target platform in order to compute their size. For example, `usize`
+implements `const Sized` as knowing only the type is `usize` and the target is
+`aarch64-unknown-linux-gnu` then we can know the size is eight bytes, and
+likewise with `armv7-unknown-linux-gnueabi` and a size of four bytes.
+
+Some types' size can be computed using only knowledge of the type and target
+platform, but only at runtime, and these types implement `Sized` (notably,
+not `const`). For example, `svint8_t` is a scalable vector type (from
+[rfc#3268][rfc_scalable_vectors]) whose length depends on the specific
+implementation of the target (i.e. it may be 128 bits on one processor and
+256 bits on another).
+
+`const ValueSized` requires more knowledge than `const Sized` to compute the size:
+it may additionally require a value  (therefore `size_of` is not implemented for
+`ValueSized`, only `size_of_val`). For example, `[usize]` implements
+`const ValueSized` as knowing the type and target is not sufficient, the number of
+elements in the slice must also be known, which requires having the value.
+
+As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
+requires a value, knowledge of the type and target platform, and can only be
+computed at runtime. For example, `[svint8_t]` requires a value to know how
+many elements there are, and then information from the runtime environment
+to know the size of a `svint8_t`.
+
+`Pointee` is implemented by any type that can be used behind a pointer, which is
+to say, every type (put otherwise, these types may or may not be sized at all).
+For example, `Pointee` is therefore implemented on a `u32` which is trivially
+sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
+runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
+no known size.
+
+All type parameters have an implicit bound of `const Sized` which will be
+automatically removed if a `~const Sized`, `Sized`, `const ValueSized`,
+`~const ValueSized`, `ValueSized` or `Pointee` bound is present instead.
+
+Prior to the introduction of `ValueSized` and `Pointee`, `Sized`'s implicit bound
+(now a `const Sized` implicit bound) could be removed using the `?Sized` syntax,
+which is now equivalent to a `ValueSized` bound in non-`const fn`s and
```

Comment:
this reminds me of C++ classes where you can figure out the size of the dynamic type by reading the vtable pointer, but, unlike Rust, that requires dereferencing the data part of the pointer to the type, whereas in Rust that info is passed as the pointer's metadata. So maybe we need both `ValueSized` and `RefSized` where for `RefSized` you can pass any old pointer to the type (since the metadata of a pointer must be valid but the data pointer doesn't need to), but for `ValueSized` you have to be able to dereference the data pointer, so takes something like `&T` but without the aliasing guarantees.

Reply:
> I don't think this is true, specifically `Mutex<T: ValueSized>` cannot be `ValueSized` as conceptually it would require locking the mutex in order to call `size_of_val` on the wrapped type. That feels terrifying and it's currently observable that's not the case because it doesn't deadlock when you call `size_of_val` while holding the Mutex.

Could you give the full code example of what you have in mind?  I want to check that what I have in mind here exactly matches what you do.

Reply:
This is a tad conceptual as it requires types that don't exist in rust as it exists today, but would exist with something like custom DSTs. Imagine this:
```rust
// Magic syntax that means that CStr isn't Sized
struct CStr(..);

impl ValueSized for CStr {
    // Not being proposed here but could exist
    // with custom DSTs.
    fn size_of_val(&self) -> usize {
        let data = self as *const u8;
        // Find first null byte and return length
    }
}
```
As you can see this type has to inspect the data behind the `&self` pointer in order to determine its size. (Admittedly I'm not entirely convinced we'd ever want this in rust because it feels like a performance foot gun, but this matches the semantics of `ValueSized` as described in the RFC)

Now consider `size_of_val(&Mutex<CStr>)` the only way that function can execute is if the `size_of_val` implementation for Mutex acquires a lock on the inner data, this is bad and it's observably not the case because this code doesn't deadlock: 
```rust
let mutex = Mutex::new(7);
let _guard = mutex.lock().unwrap();
size_of_val(&mutex);
```

The issue is that `ValueSized` as described doesn't match current rust's rules, where a type is only allowed to access the pointer metadata in order to determine its size. Changing the semantics to be that would mean that `CStr` couldn't implement `ValueSized`, meaning that `size_of_val(&Mutex<CStr>)` would throw a compile error as `Mutex<CStr>` also wouldn't implement `ValueSized`.

Reply:
I don't think this interaction with mutexes and `size_of_val` would be that surprising.

If you have a type that you know requires a value to compute its size then computing the size of a mutex containing it would require the mutex lock itself. If we had reason to believe it was common for generic code that always works today to start deadlocking when instantiated with a `ValueSized` type, then that would give me pause, but otherwise I don't think this is too bad, it's just the natural interaction of mutexes and value-sized types. That said, I do think that `MetaSized` is a perfectly acceptable alternative that captures everything that we want from `ValueSized` (to the best of my knowledge at least).

I've added an alternative that describes how `MetaSized` could be included in this RFC instead of `ValueSized` if this were an issue we wanted to avoid.

Reply:
I don't think the mutex example highlights the importance of `MetaSized` enough if we ever want to allow custom code in `size_of_val` (such as locking a mutex).

On stable rust it is currently possible to call `size_of_val` on `&UnsafeCell<dyn Trait>`, so this must continue to compile.

If we were to define `T: ValueSized` as "the size of `T` can be determined from a value of type `&T`", then there is no way to express the (compiler-builtin) impl of `ValueSized` for `UnsafeCell` without `MetaSized`.

For most normal types this impl would be:
```rust
impl const ValueSized for Foo where tail(Foo): ~const ValueSized {}
```

However, for `UnsafeCell` this would be incorrect:
```rust
impl<T> const ValueSized for UnsafeCell<T> // the size of UnsafeCell<T> can be known from a value of type &UnsafeCell<T> ...
where T: ~const ValueSized // ... if the size of T can be known from a value of type &T
{}
```

The size of `UnsafeCell<T>` **cannot** be determined by a value of type `&UnsafeCell<T>` if computing the size from `&T` requires to run user code, because there is no way to safely obtain a reference to the inner type.

Instead, we need some other trait bound to correctly express under what circumstances we can obtain the size of an `UnsafeCell<T>` if we have a `&UnsafeCell<T>` and maintain backwards compatibility with stable rust -- this trait bound is exactly `T: MetaSized`! 
```rust
impl<T> const ValueSized for UnsafeCell<T> where T: ~const MetaSized {} // OK
```

Reply:
I've added more to this alternative's section based on your `UnsafeCell` example.